ui/nuxeo-document-layout.js

```diff
@@ -0,0 +1,305 @@
+        },
+        /**
+         * The base url for the layout href.
+         */
+        hrefBase: {
+          type: String,
+          value: '',
+        },
+        /**
+         * The function used to build the layout href, using `hrefTemplate` as the function body.
+         */
+        hrefFunction: {
+          type: Function,
+          computed: '_buildHrefFn(hrefTemplate)',
+        },
+        _model: {
+          type: Object,
+          value: {},
+          readOnly: true,
+        },
+        _href: {
+          type: String,
+          readOnly: true,
+        },
+        _errorMessages: {
+          type: Array,
+          readOnly: true,
+          value: [],
+        },
+      };
+    }
+
+    static get observers() {
+      return ['_loadLayout(document, layout, hrefFunction, hrefBase)'];
+    }
+
+    /**
+     * Fired on the next render after the document layout changes.
+     *
+     * @event document-layout-changed
+     * @param {string} layout The document layout.
+     * @param {object} element The document layout element.
+     */
+
+    /**
+     * Returns the layout element.
+     */
+    get element() {
+      return this.$.layout.element;
+    }
+
+    /**
+     * Runs form validation on the document layout.
+     */
+    validate() {
+      return this.$.layout.validate();
+    }
+
+    /**
+     * Applies focus on the first layout child with the `autofocus` attribute.
+     *
+     * Note: This method won't do anything if there's already a child element focused.
+     */
+    applyAutoFocus() {
+      const focusableElement = this._getFocusableElement(this.element);
+      if (focusableElement) {
+        focusableElement.focus();
+      }
+    }
+
+    /**
+     * Displays any errors present in a validation report, and invalidates layout widgets bound to document
+     * fields whose constraints were violated.
+     * @param {object} validationReport The validation report to be displayed.
+     */
+    reportValidation(validationReport) {
+      this._resetValidationErrors();
+      validationReport.violations.reverse().forEach((violation) => {
+        this.invalid = true;
+        if (violation.path) {
+          violation.path.forEach((p) => {
+            const widgets = this._getBoundElements(`document.properties.${p.field_name}`);
+            if (widgets) {
+              const msg = this.i18n(violation.messageKey, violation.invalid_value, p.field_name);
+              if (msg === violation.messageKey && violation.constraint && violation.constraint.name) {
+                this._addValidationError(
+                  this.i18n(
+                    `label.schema.constraint.violation.${violation.constraint.name}`,
+                    violation.invalid_value,
+                    p.field_name,
+                    ...Object.values(violation.constraint.parameters),
+                  ),
+                );
+              } else {
+                this._addValidationError(msg);
+              }
```

Comment:
If I understood correctly, this specific part is not being tested, but no rush to do it.

Reply:
This is tested by the global violation test case.